ToolsPanel: Improve unit tests

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/35057

## Description

This PR improves the `ToolsPanel` unit tests so there are no uses of `querySelector`, `querySelectorAll`, and no accessing of parent nodes.

It also simulates a change of block selection with a SlotFill provided control and ensures the `onDeselect` callback isn't called erroneously.

## How has this been tested?

Run `npm run test-unit packages/components/src/tools-panel/test`

## Screenshots <!-- if applicable -->

<img width="1045" alt="Screen Shot 2021-10-15 at 5 02 48 pm" src="https://user-images.githubusercontent.com/60436221/137445904-6560480d-0fa7-42b5-9401-2c5636b32ad0.png">

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Remove use of querySelector from unit tests

aaronrobertshaw committed:
Do not rely on accessing parent node

ciampo commented on packages/components/src/tools-panel/test/index.js:
```diff
@@ -101,6 +101,7 @@ const renderGroupedItemsInPanel = () => {
 const WrappedItem = ( { text, ...props } ) => {
 	return (
 		<div className="wrapped-panel-item-container">
```
I think the `className` can be removed here?

aaronrobertshaw replied:
✅ Done. Thanks.

ciampo commented on packages/components/src/tools-panel/test/index.js:
```diff
@@ -166,9 +163,9 @@ describe( 'ToolsPanel', () => {
 	describe( 'basic rendering', () => {
 		it( 'should render panel', () => {
-			const { container } = renderPanel();
+			renderPanel();
```
This unit test and the following one (`should render non panel item child`) are now testing for the same thing — maybe a better test here would be to make sure that all the essential pieces of the UI are rendered?

```diff
diff --git a/packages/components/src/tools-panel/test/index.js b/packages/components/src/tools-panel/test/index.js
index b4be39e102..df265a8a3e 100644
--- a/packages/components/src/tools-panel/test/index.js
+++ b/packages/components/src/tools-panel/test/index.js
@@ -165,7 +165,15 @@ describe( 'ToolsPanel', () => {
 		it( 'should render panel', () => {
 			renderPanel();
 
-			expect( screen.getByText( 'Visible' ) ).toBeInTheDocument();
+			expect(
+				screen.getByText( defaultProps.label )
+			).toBeInTheDocument();
+			expect(
+				screen.getByRole( 'button', {
+					name: /view([\w\s]+)options/i,
+				} )
+			).toBeInTheDocument();
+			expect( screen.getByText( 'Example control' ) ).toBeInTheDocument();
 		} );
 
 		it( 'should render non panel item child', () => {
```

aaronrobertshaw replied:
Silly oversight on my behalf. Thanks for catching this. I've made this initial test more generalised to cover all the main parts of the panel.

ciampo commented on packages/components/src/tools-panel/test/index.js:
```diff
@@ -166,9 +163,9 @@ describe( 'ToolsPanel', () => {
+			expect( screen.getByText( 'Visible' ) ).toBeInTheDocument();
 		} );
 
 		it( 'should render non panel item child', () => {
```
Also another small nit: in this test, the line

```js
const nonPanelItem = screen.queryByText( 'Visible' );
```

can be changed to

```js
const nonPanelItem = screen.getByText( 'Visible' );
```

since we're expecting that element to be in the document.

aaronrobertshaw replied:
This test was rolled into the updated first test as part of the more generalised initial check that the panel renders.

I did find one other case where I'd left a `queryByText` after a refactor/change of behaviour so I've fixed that up. Thanks for the reminder to check that.

ciampo commented on packages/components/src/tools-panel/test/index.js:
```diff
@@ -316,16 +313,14 @@ describe( 'ToolsPanel', () => {
-			const menuGroups = container.querySelectorAll(
-				'.components-menu-group'
-			);
+			const menuGroups = screen.getAllByRole( 'group' );
```
Neat!

ciampo commented on packages/components/src/tools-panel/test/index.js:
```diff
@@ -337,15 +332,12 @@ describe( 'ToolsPanel', () => {
-			// all the item's children. So we should still find the container
-			// element but not the text etc within.
+			// all the item's children. So the container element will still be
+			// there holding it's position but the inner text etc should not be
```
typo

```suggestion
			// there holding its position but the inner text etc should not be
```

ciampo reviewed (commented):
Thank you @aaronrobertshaw for working on this! I left a few inline comments.

> No test for the above case has been added yet as the issue stems from the switching between blocks. This switching of block selections results in the fills rendered in a single panel being mutated. I haven't found a means of triggering this via the unit tests.

@andrewserong do you have any suggestions on how we could check for this behaviour?

aaronrobertshaw committed:
Remove class from wrapped panel items in tests

aaronrobertshaw committed:
Merge tests to single general rendering check

Now we aren't checking specifically just for the panel container's existence there is overlap with other tests and the benefit of separating individual concerns is lost. For a general overview test it seems ok to merge them.

aaronrobertshaw committed:
Switch queryByText to getByText

After an earlier refactor it was missed that this could be the more direct getByText call instead.

aaronrobertshaw committed:
Fix typo

aaronrobertshaw committed:
Add a test to cover the check against panelId for callbacks.

aaronrobertshaw commented:
> No test for the above case has been added yet as the issue stems from the switching between blocks. This switching of block selections results in the fills rendered in a single panel being mutated. I haven't found a means of triggering this via the unit tests.

> andrewserong do you have any suggestions on how we could check for this behaviour?

I've revisited this with some fresh eyes after the weekend. The best I could find to test that the `onDeselect` callback isn't called when block selection switches was to:

1. Create a contrived `ToolsPanelContext` 
2. Render a `ToolsPanelItem` within that context (allows for the item to set its internal state that it was previously selected)
3. Update the `ToolsPanelContext` to switch the `panelId` and set in the `menuItems` state the item was not selected
4. Rerender the `ToolsPanelItem`
5. Assert that the item's `onDeselect` callback isn't called

Removing the check added in https://github.com/WordPress/gutenberg/pull/35375 or fudging the `panelId` back to the original value causes the test to fail for me as expected.

aaronrobertshaw committed:
Add link to PR/issue that the previous test covers

andrewserong reviewed (approved):
Thanks for digging into the issue with changing panelIds @aaronrobertshaw, much appreciated! I'm glad you managed to come up with an example that mimics the behaviour we observed when switching blocks, without needing to pull in the actual block switching behaviour 👍

It's a complex situation to deal with, but the inline comments make sense to me, and for someone glancing at the test, even if they don't understand all the moving pieces, it's clear that the main thing that's changing between the two renders is that the panelId has changed, along with the optional item being toggled. While it is testing an internal state, in this situation, I think it's worth it to include the regression test how you've written it, without having to artificially build out a block switching UI for the purposes of the test. But most importantly, we've now got test coverage for a difficult to test edge case! 🎉

The changes to the existing tests all look good to me, and I confirmed that reverting https://github.com/WordPress/gutenberg/pull/35375 causes the additional test to fail 👍
